### PR TITLE
Movement pruning

### PR DESCRIPTION
**Pull request by VictorSanh**
This PR adds the code to reproduce the results of our recent work on Movement pruning.
Some supplemental treats:
- A notebook showcasing how to efficiently store an extremely sparse model
- Sharing a couple of fine-pruned checkpoints (PruneBERT)
- Details of all hyper-parameters and results

### CODE REVIEW

**Comment by VictorSanh** [comment]
Thanks for the valuable inputs on quantization @mfuntowicz! --> `Saving_PruneBERT.ipynb`

File: examples/movement-pruning/README.md [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+</div>
+
+## Extreme sparsity and efficient storage
+
+One promise of extreme pruning is to obtain extremely small models that can be easily sent (and stored) on edge devices. By setting weights to 0., we remove part of the information we need to store, and thus decreasing the memory size. We are able to obtain extremely sparse fine-pruned models with movement pruning: ~95% of the dense performance with ~5% of total remaining weights in the BERT encoder.
+
+In [this notebook](https://github.com/huggingface/transformers/blob/master/examples/movement-pruning/Saving_PruneBERT.ipynb), we showcase how we can leverage standard tools that exist out-of-the-box to efficiently store an extremely sparse question answering model (only 6% of total remaining weights in the encoder). We are able to reduce the memory size of the encoder **from the 340MB (the orignal dense BERT) to 11MB**, without any additional training of the model (every operation is *post fine-pruning*). It is sufficiently small to store it on a [91' floppy disk](https://en.wikipedia.org/wiki/Floptical) 📎!
```
Comment (thomwolf) [inline examples/movement-pruning/README.md:24]: Well a floppy disk was 1.4MB so we are not there yet. But not too far as well :-)

File: examples/movement-pruning/README.md [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+We detail below how to reproduce the results reported in the paper. We use SQuAD as a running example. Commands (and scripts) can be easily adapted for other tasks.
+
+The following command fine-prunes a pre-trained `BERT-base` on SQuAD using movement pruning towards 10% of remaining weights (90% sparsity). Note that we freeze all the embeddings modules (from the pre-trained value) and only prune the Fully Connected layers.
+
+```bash
+SERIALIZAION_DIR=<OUTPUT_DIR>
+SQUAD_DATA=<SQUAD_DATA>
+
+python examples/emmental/masked_run_squad.py \
```
Comment (thomwolf) [inline examples/movement-pruning/README.md:53]: Rename `emmental`

File: examples/movement-pruning/README.md [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+```bash
+python examples/movement-pruning/bertarize.py \
+    --pruning_method sigmoied_threshold \
+    --threshold 0.1 \
+    --model_name_or_path $SERIALIZAION_DIR
+```
+
+## Hyper-parameters
+
+For reproducibility purposes, we share the detailed results presented in the paper. This [spreadsheet](TODO) exhaustively describes the individual hyper-parameters used for each data point.
```
Comment (thomwolf) [inline examples/movement-pruning/README.md:160]: Don't forget the `TODO`

**Review by thomwolf — APPROVED** [review]
Look good to me!

**Review by julien-c — APPROVED** [review]

**Review by LysandreJik — APPROVED** [review]
This is great. Didn't you want to include the `MaskedBertXXX` directly in the library, as was done with `DistilBERT`?

We can also do it at a later date, as we'll need to write the tests (kudos on the docs, looks nice!)

**Review by joeddav — APPROVED** [review]
Looks great. Well organized & clearly communicated.

Would it be worthwhile to add the released prunebert encoders to exBERT? I don't think we have any fine-tuned models there right now but it'd be cool to let people see how the movement pruning process affects the attention distributions.

File: examples/movement-pruning/README.md [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+## Remarks
+
+Early experiments show that even though models fine-pruned with (soft) movement pruning are extremely sparse, they do not benefit from significant improvement in terms of inference speed when using the standard PyTorch inference.
+We are currently benchmarking and exploring inference setups specifically for sparse architectures.
```
Comment (julien-c) [inline examples/movement-pruning/README.md:165]:
```suggestion
We are currently benchmarking and exploring inference setups specifically for sparse architectures.
In particular, hardware manufacturers are announcing devices that will speedup inference for sparse networks considerably.
```

File: examples/movement-pruning/README.md [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    --pruning_method sigmoied_threshold \
+    --threshold 0.1 \
+    --model_name_or_path $SERIALIZAION_DIR
+```
+
+## Hyper-parameters
+
+For reproducibility purposes, we share the detailed results presented in the paper. This [spreadsheet](TODO) exhaustively describes the individual hyper-parameters used for each data point.
+
+## Remarks
```
Comment (julien-c) [inline examples/movement-pruning/README.md:162]:
```suggestion
## Inference speed
```

File: examples/movement-pruning/README.md (hunk 1 above)
Comment (VictorSanh) [reply examples/movement-pruning/README.md:24]: According to [Wikipedia](https://en.wikipedia.org/wiki/Floppy_disk) (i was either too young or not born yet at this time), there were a couple of developments after that. The LS Floptical has a 21MB storage capacity.

File: examples/movement-pruning/README.md (hunk 2 above)
Comment (VictorSanh) [reply examples/movement-pruning/README.md:53]: yes indeed

File: examples/movement-pruning/README.md (hunk 3 above)
Comment (VictorSanh) [reply examples/movement-pruning/README.md:160]: 👍

**Comment by VictorSanh** [comment]
> This is great. Didn't you want to include the `MaskedBertXXX` directly in the library, as was done with `DistilBERT`?
> 
> We can also do it at a later date, as we'll need to write the tests (kudos on the docs, looks nice!)

Good question. I think it's fair for the moment to leave it outside of the library itself: once a pre-trained model has been fine-pruned, it can be pruned once for all and you end up with a standard `BertForSequenceClassification` for instance. So I see it more as an "intermediate tool".

**Comment by VictorSanh** [comment]
> Looks great. Well organized & clearly communicated.
> 
> Would it be worthwhile to add the released prunebert encoders to exBERT? I don't think we have any fine-tuned models there right now but it'd be cool to let people see how the movement pruning process affects the attention distributions.

Good point! I'll have a look.

**Pull request merged**